File: backend/api/routers/blog.py
from fastapi import APIRouter, HTTPException
from typing import List, Optional
import json
from pathlib import Path
from datetime import datetime

router = APIRouter()

# Path to blog data
BLOG_DATA_PATH = Path(__file__).parent.parent.parent / "data" / "blog-articles.json"

# Cache for blog data
blog_cache = None

def load_blog_data():
    """Load blog data from JSON file."""
    global blog_cache
    if blog_cache is None and BLOG_DATA_PATH.exists():
        with open(BLOG_DATA_PATH, 'r', encoding='utf-8') as f:
            blog_cache = json.load(f)
    return blog_cache or {"posts": []}

@router.get("/posts")
async def get_blog_posts(
    page: int = 1,
    limit: int = 10,
    search: Optional[str] = None
):
    """Get paginated list of blog posts."""
    data = load_blog_data()
    posts = data.get("posts", [])
    
    # Search functionality
    if search:
        search_lower = search.lower()
        posts = [
            post for post in posts
            if search_lower in post.get("title", "").lower() or
               search_lower in post.get("excerpt", "").lower()
        ]
    
    # Pagination
    start = (page - 1) * limit
    end = start + limit
    paginated_posts = posts[start:end]
    
    # Return only necessary fields for listing
    return {
        "posts": [
            {
                "id": post["id"],
                "title": post["title"],
                "excerpt": post["excerpt"],
                "date": post["date"],
                "slug": post["slug"],
                "readTime": post["readTime"]
            }
            for post in paginated_posts
        ],
        "total": len(posts),
        "page": page,
        "pages": (len(posts) + limit - 1) // limit
    }

@router.get("/posts/{slug}")
async def get_blog_post(slug: str):
    """Get a single blog post by slug."""
    data = load_blog_data()
    posts = data.get("posts", [])
    
    # Find post by slug
    post = next((p for p in posts if p.get("slug") == slug), None)
    
    if not post:
        raise HTTPException(status_code=404, detail="Blog post not found")
    
    return post

@router.get("/posts/related/{slug}")
async def get_related_posts(slug: str, limit: int = 5):
    """Get related blog posts based on current post."""
    data = load_blog_data()
    posts = data.get("posts", [])
    
    # Find current post
    current_post = next((p for p in posts if p.get("slug") == slug), None)
    if not current_post:
        raise HTTPException(status_code=404, detail="Blog post not found")
    
    # Simple related posts: just get other posts excluding current one
    # In a real implementation, you'd use tags, categories, or ML to find related content
    related = [p for p in posts if p.get("slug") != slug][:limit]
    
    return {
        "posts": [
            {
                "id": post["id"],
                "title": post["title"],
                "excerpt": post["excerpt"],
                "slug": post["slug"],
                "readTime": post["readTime"]
            }
            for post in related
        ]
    }

@router.get("/categories")
async def get_blog_categories():
    """Get list of blog categories."""
    # For now, return predefined categories
    # In a real implementation, these would be extracted from posts
    return {
        "categories": [
            {"id": "relationships", "name": "Relationships", "count": 45},
            {"id": "body-language", "name": "Body Language", "count": 32},
            {"id": "communication", "name": "Communication", "count": 28},
            {"id": "dating", "name": "Dating", "count": 34}
        ]
    }

File: personality-spark-api/src/routes/blog.ts
import { Hono } from 'hono';
import { zValidator } from '@hono/zod-validator';
import { z } from 'zod';
import type { Context } from '../types/env';

export const blogRouter = new Hono<Context>();

// Query parameter schemas
const PaginationSchema = z.object({
  page: z.string().regex(/^\d+$/).transform(Number).pipe(z.number().min(1).max(1000)).optional().default('1'),
  limit: z.string().regex(/^\d+$/).transform(Number).pipe(z.number().min(1).max(2000)).optional().default('20'),
  search: z.string().max(100).optional()
});

const RelatedPostsSchema = z.object({
  limit: z.string().regex(/^\d+$/).transform(Number).pipe(z.number().min(1).max(20)).optional().default('5')
});

// Load blog data from chunked files or KV
async function loadBlogData(c: any) {
  try {
    // Try to get blog data from KV storage first
    const cached = await c.env.CACHE.get('blog-data-all', 'json');
    if (cached && cached.posts && cached.posts.length > 0) {
      return cached;
    }
  } catch (error) {
    console.error('Error loading from KV:', error);
  }
  
  try {
    // First load the index to know about chunks
    const url = new URL(c.req.url);
    const indexUrl = `${url.protocol}//${url.host}/blog-index.json`;
    const indexResponse = await fetch(indexUrl);
    
    if (!indexResponse.ok) {
      throw new Error('Failed to fetch blog index');
    }
    
    const indexData = await indexResponse.json();
    
    // Load all chunks in parallel
    const chunkPromises = indexData.chunks.map(async (chunk: any) => {
      const chunkUrl = `${url.protocol}//${url.host}/${chunk.file}`;
      const response = await fetch(chunkUrl);
      if (!response.ok) {
        throw new Error(`Failed to fetch ${chunk.file}`);
      }
      return response.json();
    });
    
    const chunks = await Promise.all(chunkPromises);
    
    // Combine all posts from chunks
    const allPosts = chunks.flatMap(chunk => chunk.posts);
    
    const data = { posts: allPosts };
    
    // Cache the combined data in KV for future use
    try {
      await c.env.CACHE.put('blog-data-all', JSON.stringify(data), {
        expirationTtl: 3600 // Cache for 1 hour
      });
    } catch (cacheError) {
      console.error('Error caching blog data:', cacheError);
    }
    
    return data;
  } catch (error) {
    console.error('Error loading blog data from chunks:', error);
    
    // Fallback to original blog-data.json if chunks fail
    try {
      const url = new URL(c.req.url);
      const blogDataUrl = `${url.protocol}//${url.host}/blog-data.json`;
      const response = await fetch(blogDataUrl);
      
      if (response.ok) {
        const data = await response.json();
        return data;
      }
    } catch (fallbackError) {
      console.error('Error fetching fallback blog data:', fallbackError);
    }
  }
  
  // Return empty data as last resort
  return { posts: [] };
}

// Get all blog posts
blogRouter.get('/posts', 
  zValidator('query', PaginationSchema),
  async (c) => {
    const { page, limit, search } = c.req.valid('query');
    
    const data = await loadBlogData(c);
    let posts = data.posts || [];
    
    // Search functionality with sanitized input
    if (search) {
      const searchLower = search.toLowerCase();
      posts = posts.filter((post: any) =>
        post.title.toLowerCase().includes(searchLower) ||
        post.excerpt.toLowerCase().includes(searchLower)
      );
    }
    
    // Pagination with validated bounds
    const start = (page - 1) * limit;
    const end = start + limit;
    const paginatedPosts = posts.slice(start, end);
    
    return c.json({
      posts: paginatedPosts.map((post: any) => ({
        id: post.id,
        title: post.title,
        excerpt: post.excerpt,
        date: post.date,
        slug: post.slug,
        readTime: post.readTime
      })),
      total: posts.length,
      page: page,
      pages: Math.ceil(posts.length / limit)
    });
  }
);

// Get a single blog post by slug
blogRouter.get('/posts/:slug', async (c) => {
  const slug = c.req.param('slug');
  
  // Validate slug format - alphanumeric with hyphens only
  const slugRegex = /^[a-z0-9]+(?:-[a-z0-9]+)*$/;
  if (!slugRegex.test(slug) || slug.length > 100) {
    return c.json({ 
      error: 'Invalid slug format',
      message: 'Slug must contain only lowercase letters, numbers, and hyphens'
    }, 400);
  }
  
  const data = await loadBlogData(c);
  const posts = data.posts || [];
  
  const post = posts.find((p: any) => p.slug === slug);
  
  if (!post) {
    return c.json({ error: 'Blog post not found' }, 404);
  }
  
  // Content is already HTML, no need to convert
  return c.json(post);
});

// Get related posts
blogRouter.get('/posts/:slug/related', 
  zValidator('query', RelatedPostsSchema),
  async (c) => {
    const slug = c.req.param('slug');
    const { limit } = c.req.valid('query');
    
    // Validate slug format
    const slugRegex = /^[a-z0-9]+(?:-[a-z0-9]+)*$/;
    if (!slugRegex.test(slug) || slug.length > 100) {
      return c.json({ 
        error: 'Invalid slug format',
        message: 'Slug must contain only lowercase letters, numbers, and hyphens'
      }, 400);
    }
    
    const data = await loadBlogData(c);
    const posts = data.posts || [];
    
    // Find current post
    const currentPost = posts.find((p: any) => p.slug === slug);
    if (!currentPost) {
      return c.json({ error: 'Blog post not found' }, 404);
    }
    
    // Get related posts (excluding current)
    const related = posts
      .filter((p: any) => p.slug !== slug)
      .slice(0, limit);
    
    return c.json({
      posts: related.map((post: any) => ({
        id: post.id,
        title: post.title,
        excerpt: post.excerpt,
        slug: post.slug,
        readTime: post.readTime
      }))
    });
  }
);

// Get blog categories
blogRouter.get('/categories', async (c) => {
  // Return predefined categories for now
  return c.json({
    categories: [
      { id: 'relationships', name: 'Relationships', count: 45 },
      { id: 'body-language', name: 'Body Language', count: 32 },
      { id: 'communication', name: 'Communication', count: 28 },
      { id: 'dating', name: 'Dating', count: 34 }
    ]
  });
});